extension/inet/inet_functions.cpp

```diff
@@ -3,13 +3,65 @@
 #include "duckdb/common/string_util.hpp"
 #include "duckdb/common/pair.hpp"
 #include "duckdb/common/operator/cast_operators.hpp"
+#include "duckdb/common/operator/subtract.hpp"
+#include "duckdb/common/operator/add.hpp"
 #include "duckdb/common/types/cast_helpers.hpp"
 #include "duckdb/common/vector_operations/generic_executor.hpp"
 
+/* warning suppress begin */
+/* Only a few functions out of numeric_cast are used*/
+#if defined(__clang__)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-function"
+#elif defined(__GNUC__)
+#if (__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
+#pragma GCC diagnostic push
+#endif
+#pragma GCC diagnostic ignored "-Wunused-function"
+#endif
+
+#include "duckdb/common/operator/numeric_cast.hpp"
+
+#if defined(__clang__)
+#pragma clang diagnostic pop
+#elif defined(__GNUC__)
+#if (__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
+#pragma GCC diagnostic pop
+#endif
+#endif /* warning suppress end */
+
 namespace duckdb {
 
+// While the address field is better represented as a uhugeint_t, the original
+// implementation used hugeint_t, so to maintain backward-compatibility it will
+// continue to be stored as signed. However, operations on the address values
+// will use the unsigned variant, so use the functions below to convert to/from
+// the compatible representation.
 using INET_TYPE = StructTypeTernary<uint8_t, hugeint_t, uint16_t>;
 
+static uhugeint_t from_compat_addr(hugeint_t compat_addr, IPAddressType addr_type) {
+	uhugeint_t retval = static_cast<uhugeint_t>(compat_addr);
+	// Only flip the bit for order on IPv6 addresses. It can never be set in IPv4
+	if (addr_type == IPAddressType::IP_ADDRESS_V6) {
+		// The top bit is flipped when storing as the signed hugeint so that sorting
+		// works correctly. Flip it back here to have a proper unsigned value.
+		return retval ^ (uhugeint_t(1) << 127);
```

Comment:
Bitwise ops on `uhugeint` are somewhat expensive - perhaps we can do the bitwise operation directly on the component that is affected instead (e.g. `new_addr.upper` or `new_addr.lower`)?

Reply:
I've updated the implementation to perform bit ops directly on the two halves of the hugeint object.